File: Sources/TemperatureReader.m
#import "TemperatureReader.h"

#include <Foundation/Foundation.h>
#include <IOKit/hidsystem/IOHIDEventSystemClient.h>

typedef struct __IOHIDEvent *IOHIDEventRef;
typedef struct __IOHIDServiceClient *IOHIDServiceClientRef;
#ifdef __LP64__
typedef double IOHIDFloat;
#else
typedef float IOHIDFloat;
#endif

IOHIDEventSystemClientRef IOHIDEventSystemClientCreate(CFAllocatorRef allocator);
int IOHIDEventSystemClientSetMatching(IOHIDEventSystemClientRef client, CFDictionaryRef match);
IOHIDEventRef IOHIDServiceClientCopyEvent(IOHIDServiceClientRef service, int64_t type, int32_t options, int64_t timestamp);
IOHIDFloat IOHIDEventGetFloatValue(IOHIDEventRef event, int32_t field);
CFArrayRef IOHIDEventSystemClientCopyServices(IOHIDEventSystemClientRef client);


#define IOHIDEventFieldBase(type) (type << 16)
#define kIOHIDEventTypeTemperature 15

@implementation TemperatureReading
- (instancetype)initWithName:(NSString *)name value:(double)value {
    self = [super init];
    if (self) {
        _name = [name copy];
        _value = value;
    }
    return self;
}
@end

@implementation TemperatureSnapshot
- (instancetype)initWithCPUReadings:(NSArray<TemperatureReading *> *)cpuReadings
                        gpuReadings:(NSArray<TemperatureReading *> *)gpuReadings
                        allReadings:(NSArray<TemperatureReading *> *)allReadings
                    thermalPressure:(NSString *)thermalPressure
                          timestamp:(NSDate *)timestamp {
    self = [super init];
    if (self) {
        _cpuReadings = cpuReadings;
        _gpuReadings = gpuReadings;
        _allReadings = allReadings;
        _thermalPressure = [thermalPressure copy];
        _timestamp = timestamp;
    }
    return self;
}
@end

static CFDictionaryRef Matching(int page, int usage) {
    CFStringRef keys[2];
    CFNumberRef nums[2];

    keys[0] = CFSTR("PrimaryUsagePage");
    keys[1] = CFSTR("PrimaryUsage");
    nums[0] = CFNumberCreate(kCFAllocatorDefault, kCFNumberSInt32Type, &page);
    nums[1] = CFNumberCreate(kCFAllocatorDefault, kCFNumberSInt32Type, &usage);

    CFDictionaryRef dict = CFDictionaryCreate(kCFAllocatorDefault,
                                              (const void **)keys,
                                              (const void **)nums,
                                              2,
                                              &kCFTypeDictionaryKeyCallBacks,
                                              &kCFTypeDictionaryValueCallBacks);
    CFRelease(nums[0]);
    CFRelease(nums[1]);
    return dict;
}

static NSArray<NSString *> *GetThermalNames(void) {
    CFDictionaryRef sensors = Matching(0xff00, 5);
    IOHIDEventSystemClientRef system = IOHIDEventSystemClientCreate(kCFAllocatorDefault);
    IOHIDEventSystemClientSetMatching(system, sensors);
    CFArrayRef services = IOHIDEventSystemClientCopyServices(system);

    NSMutableArray<NSString *> *names = [NSMutableArray array];
    CFIndex count = CFArrayGetCount(services);
    for (CFIndex i = 0; i < count; i++) {
        IOHIDServiceClientRef service = (IOHIDServiceClientRef)CFArrayGetValueAtIndex(services, i);
        CFStringRef name = (CFStringRef)IOHIDServiceClientCopyProperty(service, CFSTR("Product"));
        [names addObject:name ? (__bridge_transfer NSString *)name : @"noname"];
    }

    CFRelease(services);
    CFRelease(system);
    CFRelease(sensors);
    return names;
}

static NSArray<NSNumber *> *GetThermalValues(void) {
    CFDictionaryRef sensors = Matching(0xff00, 5);
    IOHIDEventSystemClientRef system = IOHIDEventSystemClientCreate(kCFAllocatorDefault);
    IOHIDEventSystemClientSetMatching(system, sensors);
    CFArrayRef services = IOHIDEventSystemClientCopyServices(system);

    NSMutableArray<NSNumber *> *values = [NSMutableArray array];
    CFIndex count = CFArrayGetCount(services);
    for (CFIndex i = 0; i < count; i++) {
        IOHIDServiceClientRef service = (IOHIDServiceClientRef)CFArrayGetValueAtIndex(services, i);
        IOHIDEventRef event = IOHIDServiceClientCopyEvent(service, kIOHIDEventTypeTemperature, 0, 0);
        double value = 0;
        if (event != NULL) {
            value = IOHIDEventGetFloatValue(event, IOHIDEventFieldBase(kIOHIDEventTypeTemperature));
            CFRelease(event);
        }
        [values addObject:@(value)];
    }

    CFRelease(services);
    CFRelease(system);
    CFRelease(sensors);
    return values;
}

static BOOL IsValidTemperature(double value) {
    return value >= 0.0 && value < 110.0;
}

static BOOL IsLikelyCPUName(NSString *name) {
    return [name hasPrefix:@"PMU tdie"] ||
           [name hasPrefix:@"PMU tdev"] ||
           [name hasPrefix:@"pACC MTR Temp"] ||
           [name hasPrefix:@"eACC MTR Temp"] ||
           [name rangeOfString:@"CPU"].location != NSNotFound;
}

static BOOL IsExplicitLikelyGPUName(NSString *name) {
    return [name hasPrefix:@"GPU MTR Temp Sensor"] ||
           [name rangeOfString:@"GPU"].location != NSNotFound;
}

static BOOL IsFallbackLikelyGPUName(NSString *name) {
    return [name hasPrefix:@"PMU2 tdie"] || [name hasPrefix:@"PMU2 tdev"];
}

static NSString *ThermalPressureString(void) {
    switch ([NSProcessInfo processInfo].thermalState) {
        case NSProcessInfoThermalStateNominal:
            return @"Nominal";
        case NSProcessInfoThermalStateFair:
            return @"Fair";
        case NSProcessInfoThermalStateSerious:
            return @"Serious";
        case NSProcessInfoThermalStateCritical:
            return @"Critical";
    }
}

@implementation TemperatureReader

- (TemperatureSnapshot *)readSnapshot {
    NSArray<NSString *> *names = GetThermalNames();
    NSArray<NSNumber *> *values = GetThermalValues();
    NSMutableArray<TemperatureReading *> *allReadings = [NSMutableArray array];
    NSMutableArray<TemperatureReading *> *cpuReadings = [NSMutableArray array];
    NSMutableArray<TemperatureReading *> *explicitGPUReadings = [NSMutableArray array];
    NSMutableArray<TemperatureReading *> *fallbackGPUReadings = [NSMutableArray array];

    NSUInteger count = MIN(names.count, values.count);
    for (NSUInteger i = 0; i < count; i++) {
        NSString *name = names[i];
        double value = values[i].doubleValue;
        if (!IsValidTemperature(value)) {
            continue;
        }

        TemperatureReading *reading = [[TemperatureReading alloc] initWithName:name value:value];
        [allReadings addObject:reading];

        if (IsLikelyCPUName(name)) {
            [cpuReadings addObject:reading];
        }
        if (IsExplicitLikelyGPUName(name)) {
            [explicitGPUReadings addObject:reading];
        } else if (IsFallbackLikelyGPUName(name)) {
            [fallbackGPUReadings addObject:reading];
        }
    }

    NSArray<TemperatureReading *> *gpuReadings = explicitGPUReadings.count > 0 ? explicitGPUReadings : fallbackGPUReadings;
    return [[TemperatureSnapshot alloc] initWithCPUReadings:cpuReadings
                                                gpuReadings:gpuReadings
                                                allReadings:allReadings
                                            thermalPressure:ThermalPressureString()
                                                  timestamp:[NSDate date]];
}

@end

File: Sources/TemperatureReader.swift
import Foundation

final class TemperatureReader {
    func readTemperature(mode: DisplayMode) -> Double? {
        let sensors = Self.getThermalSensors()
        var hottest: Double?

        for sensor in sensors {
            guard sensor.value >= 0.0 && sensor.value < 110.0 else { continue }
            let match = switch mode {
            case .cpu: Self.isLikelyCPUName(sensor.name)
            case .gpu: Self.isLikelyGPUName(sensor.name)
            }
            if match, sensor.value > (hottest ?? 0) {
                hottest = sensor.value
            }
        }
        return hottest
    }
}

enum DisplayMode {
    case cpu
    case gpu

    var label: String {
        switch self {
        case .cpu: return "CPU"
        case .gpu: return "GPU"
        }
    }
}

private extension TemperatureReader {
    struct ThermalSensor {
        let name: String
        let value: Double
    }

    static func isLikelyCPUName(_ name: String) -> Bool {
        name.hasPrefix("PMU tdie") ||
        name.hasPrefix("PMU tdev") ||
        name.hasPrefix("pACC MTR Temp") ||
        name.hasPrefix("eACC MTR Temp") ||
        name.contains("CPU")
    }

    static func isLikelyGPUName(_ name: String) -> Bool {
        name.hasPrefix("GPU MTR Temp Sensor") ||
        name.contains("GPU") ||
        name.hasPrefix("PMU2 tdie") ||
        name.hasPrefix("PMU2 tdev")
    }
}

private enum HIDAPI {
    private static let handle = dlopen("/System/Library/Frameworks/IOKit.framework/IOKit", RTLD_LAZY)

    typealias CreateFunc = @convention(c) (CFAllocator?) -> UnsafeMutableRawPointer?
    typealias SetMatchingFunc = @convention(c) (UnsafeMutableRawPointer, CFDictionary) -> Int32
    typealias CopyServicesFunc = @convention(c) (UnsafeMutableRawPointer) -> CFArray?
    typealias CopyPropertyFunc = @convention(c) (UnsafeMutableRawPointer, CFString) -> Unmanaged<AnyObject>?
    typealias CopyEventFunc = @convention(c) (UnsafeMutableRawPointer, Int64, Int32, Int64) -> UnsafeMutableRawPointer?
    typealias GetFloatValueFunc = @convention(c) (UnsafeMutableRawPointer, Int32) -> Double

    static let create: CreateFunc? = load("IOHIDEventSystemClientCreate")
    static let setMatching: SetMatchingFunc? = load("IOHIDEventSystemClientSetMatching")
    static let copyServices: CopyServicesFunc? = load("IOHIDEventSystemClientCopyServices")
    static let copyProperty: CopyPropertyFunc? = load("IOHIDServiceClientCopyProperty")
    static let copyEvent: CopyEventFunc? = load("IOHIDServiceClientCopyEvent")
    static let getFloatValue: GetFloatValueFunc? = load("IOHIDEventGetFloatValue")

    private static func load<T>(_ name: String) -> T? {
        guard let handle, let sym = dlsym(handle, name) else { return nil }
        return unsafeBitCast(sym, to: T.self)
    }
}

private let kIOHIDEventTypeTemperature: Int64 = 15

private let thermalMatchingDict: CFDictionary = {
    var page: Int32 = 0xff00
    var usage: Int32 = 5
    let pageNum = CFNumberCreate(kCFAllocatorDefault, .sInt32Type, &page) as CFNumber
    let usageNum = CFNumberCreate(kCFAllocatorDefault, .sInt32Type, &usage) as CFNumber
    return [
        "PrimaryUsagePage" as CFString: pageNum,
        "PrimaryUsage" as CFString: usageNum
    ] as CFDictionary
}()

private func rawRelease(_ ptr: UnsafeMutableRawPointer) {
    Unmanaged<CFTypeRef>.fromOpaque(ptr).release()
}

private extension TemperatureReader {
    static func getThermalSensors() -> [ThermalSensor] {
        guard let create = HIDAPI.create,
              let setMatching = HIDAPI.setMatching,
              let copyServices = HIDAPI.copyServices,
              let copyProperty = HIDAPI.copyProperty,
              let copyEvent = HIDAPI.copyEvent,
              let getFloatValue = HIDAPI.getFloatValue
        else { return [] }

        guard let system = create(kCFAllocatorDefault) else { return [] }
        _ = setMatching(system, thermalMatchingDict)
        guard let services = copyServices(system) else { rawRelease(system); return [] }

        var sensors: [ThermalSensor] = []
        let count = CFArrayGetCount(services)
        let field = Int32(kIOHIDEventTypeTemperature << 16)
        for i in 0..<count {
            let ptr = UnsafeMutableRawPointer(mutating: CFArrayGetValueAtIndex(services, i))
            let name: String
            if let ref = copyProperty(ptr!, "Product" as CFString) {
                name = ref.takeRetainedValue() as? String ?? "noname"
            } else {
                name = "noname"
            }
            var value: Double = 0
            if let event = copyEvent(ptr!, kIOHIDEventTypeTemperature, 0, 0) {
                value = getFloatValue(event, field)
                rawRelease(event)
            }
            sensors.append(ThermalSensor(name: name, value: value))
        }
        rawRelease(Unmanaged.passUnretained(services).toOpaque())
        rawRelease(system)
        return sensors
    }
}
